[SPARK-35449][SQL] Only extract common expressions from CaseWhen values if elseValue is set

Kimahriman opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
  8. If you want to add or modify an error message, please read the guideline first:
     https://spark.apache.org/error-message-guidelines.html
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->
This PR fixes a bug with subexpression elimination for CaseWhen statements. https://github.com/apache/spark/pull/30245 added support for creating subexpressions that are present in all branches of conditional statements. However, for a statement to be in "all branches" of a CaseWhen statement, it must also be in the elseValue.


### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->
Fix a bug where a subexpression can be created and run for branches of a conditional that don't pass. This can cause issues especially with a UDF in a branch that gets executed assuming the condition is true.


### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->
Yes, fixes a potential bug where a UDF could be eagerly executed even though it might expect to have already passed some form of validation. For example:
```
val col = when($"id" < 0, myUdf($"id"))
spark.range(1).select(when(col > 0, col)).show()
```

`myUdf($"id")` is considered a subexpression and eagerly evaluated, because it is pulled out as a common expression from both executions of the when clause, but if `id >= 0` it should never actually be run.

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->
Updated existing test with new case.

Kimahriman committed:
CaseWhen needs to have an elseValue to be considered for subexpression elimination

Kimahriman commented:
@viirya

AmplabJenkins commented:
Can one of the admins verify this patch?

viirya commented:
ok to test

viirya reviewed (commented):
looks good. waiting for CI

SparkQA commented:
**[Test build #138722 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138722/testReport)** for PR 32595 at commit [`abc6150`](https://github.com/apache/spark/commit/abc6150ecb7db3fdfe0edb2e5f13d1257a1ae337).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43245/

AmplabJenkins commented:
Can one of the admins verify this patch?

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43245/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/43245/

SparkQA commented:
**[Test build #138722 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138722/testReport)** for PR 32595 at commit [`abc6150`](https://github.com/apache/spark/commit/abc6150ecb7db3fdfe0edb2e5f13d1257a1ae337).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/138722/

viirya commented:
cc @maropu @cloud-fan too

AmplabJenkins commented:
Can one of the admins verify this patch?

cloud-fan commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/SubexpressionEliminationSuite.scala:
```diff
@@ -226,6 +226,17 @@ class SubexpressionEliminationSuite extends SparkFunSuite with ExpressionEvalHel
     equivalence3.addExprTree(caseWhenExpr3)
     assert(equivalence3.getAllEquivalentExprs.count(_.size == 2) == 0)
+
+    val conditions4 = (GreaterThan(add1, Literal(3)), add1) ::
```
`conditions4` is not used anywhere.

Kimahriman replied:
Copy pasta, fixed the below line

cloud-fan commented:
is it a perf-only issue?

viirya commented:
> is it a perf-only issue?

In most cases, it is. But what I understand from @Kimahriman's description, seems in their usage like `when(..., udf(col))`, `udf(col)` will be wrongly treated as sub-expr and run in advance. But the udf fails for the input because it doesn't satisfy the condition.

So it could cause query error.

Kimahriman commented:
> is it a perf-only issue?

Yeah because of the UDF issue I'd consider it more a bug with performance side-effects. Whether those side-effects are positive or negative largely depends on whether https://github.com/apache/spark/pull/32559 is merged. Without it, this can increase performance by reducing the cases where you could have unused subexpressions generated. With it, it can decrease performance by not being able to create subexpressions for simple when clauses like `when(length(col) > 0, col)`

SparkQA commented:
**[Test build #138746 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138746/testReport)** for PR 32595 at commit [`a8b9e89`](https://github.com/apache/spark/commit/a8b9e89e948d0c43db18f7df9eb14af42b7582ea).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43268/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43268/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/43268/

SparkQA commented:
**[Test build #138746 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138746/testReport)** for PR 32595 at commit [`a8b9e89`](https://github.com/apache/spark/commit/a8b9e89e948d0c43db18f7df9eb14af42b7582ea).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/138746/

HyukjinKwon commented:
@Kimahriman would you mind describing what behaviour change (bug fix) happens in "Does this PR introduce any user-facing change?"? The fix itself looks making sense but it would be great to clarify what bug is fixed by this too.

Kimahriman commented:
> @Kimahriman would you mind describing what behaviour change (bug fix) happens in "Does this PR introduce any user-facing change?"? The fix itself looks making sense but it would be great to clarify what bug is fixed by this too.

Updated, sorry I never know what to put for that section for bug fixes.

maropu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala:
```diff
@@ -128,7 +128,11 @@ class EquivalentExpressions {
       // condition is true, it will be evaluated.
       val conditions = c.branches.tail.map(_._1)
-      val values = c.branches.map(_._2) ++ c.elseValue
+      val values = if (c.elseValue.nonEmpty) {
```
> However, for a statement to be in "all branches" of a CaseWhen statement, it must also be in the elseValue.

It's better to leave a simple comment here about why we need to check `elseValue`.

Kimahriman replied:
Added a comment in

maropu commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/SubexpressionEliminationSuite.scala:
```diff
@@ -226,6 +226,17 @@ class SubexpressionEliminationSuite extends SparkFunSuite with ExpressionEvalHel
+    equivalence4.addExprTree(caseWhenExpr4)
+
+    // `add1` is not in the elseValue, so we can't extract it from the branches
+    assert(equivalence4.getAllEquivalentExprs.count(_.size == 2) == 0)
```
Could you put this new test in a new test unit `test("SPARK-35449: ..."?

Kimahriman replied:
Moved to a new test

maropu reviewed (commented):
The fix itself looks fine. I left minor comments. btw, could you fill the description in the jira? It looks empty now: https://issues.apache.org/jira/browse/SPARK-35449

Kimahriman committed:
Add comment, create new test, and fix test after rebasing

Kimahriman commented:
> The fix itself looks fine. I left minor comments. btw, could you fill the description in the jira? It looks empty now: https://issues.apache.org/jira/browse/SPARK-35449

I updated the JIRA and rebased off master to get the latest subexpression PRs in. Had to update one test based on these changes.

SparkQA commented:
**[Test build #138849 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138849/testReport)** for PR 32595 at commit [`e024cec`](https://github.com/apache/spark/commit/e024cecaace8a70237ce4d0741196f28ed553cd2).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43371/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43371/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/43371/

SparkQA commented:
**[Test build #138850 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138850/testReport)** for PR 32595 at commit [`5f203d0`](https://github.com/apache/spark/commit/5f203d0288ea76953483e7b45c995c3027c0cbbc).

SparkQA commented:
**[Test build #138849 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138849/testReport)** for PR 32595 at commit [`e024cec`](https://github.com/apache/spark/commit/e024cecaace8a70237ce4d0741196f28ed553cd2).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/138849/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43372/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43372/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/43372/

SparkQA commented:
**[Test build #138850 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138850/testReport)** for PR 32595 at commit [`5f203d0`](https://github.com/apache/spark/commit/5f203d0288ea76953483e7b45c995c3027c0cbbc).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/138850/

viirya commented:
Thanks all! Merging to master/3.1. I guess there will be conflict on 3.1, but let me try first.

viirya commented:
This should to be fixed in branch-3.1 too. @Kimahriman Can you make a backport PR?

Also cc @dongjoon-hyun as he is release manager of 3.1.2.

viirya commented:
Ah, right, @maropu, because backport to 3.1 fails, so the script doesn't update it. I forgot to do it manually too. Let me do it now. Thanks!

Kimahriman commented:
> This should to be fixed in branch-3.1 too. @Kimahriman Can you make a backport PR?
> 
> Also cc @dongjoon-hyun as he is release manager of 3.1.2.

Created https://github.com/apache/spark/pull/32651
